File: scripts/agents/apio_monitor.go
// apio_monitor.go
// ===================
// Drop-in HTTP middleware for Go — paste into your project.
//
// Works with net/http, chi, gorilla/mux, gin, echo — any framework
// that accepts an http.Handler or http.HandlerFunc.
// Zero external dependencies — uses only the standard library.
//
// Setup:
//   1. Copy this file into your project (e.g. apio_monitor.go)
//   2. Change the package declaration below to match your package
//   3. Set env var: APIO_SDK_TOKEN=your_token
//   4. Wrap your handler (see examples below)
//   5. Run your app normally: go run .
//
// net/http:
//   http.ListenAndServe(":8080", ApioMiddleware(yourMux))
//
// chi:
//   r := chi.NewRouter()
//   r.Use(ApioMiddlewareFunc)
//   http.ListenAndServe(":8080", r)
//
// gin:
//   r := gin.Default()
//   r.Use(ApioGinMiddleware())
//
// Config (env vars):
//   APIO_SDK_TOKEN    — your service SDK token  (required)
//   APIO_BACKEND_URL  — e.g. http://localhost:4000  (default)

package main // ← change this to match your package

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"
)

// ── Config ───────────────────────────────────────────────────────────────────

var (
	apioSDKToken   = os.Getenv("APIO_SDK_TOKEN")
	apioBackendURL = strings.TrimRight(apioEnvOr("APIO_BACKEND_URL", "http://localhost:4000"), "/")
	apioIngestURL  = apioBackendURL + "/api/v1/ingest"
)

func apioEnvOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}

// ── Event ─────────────────────────────────────────────────────────────────────

type apioEvent struct {
	Method         string            `json:"method"`
	URL            string            `json:"url"`
	StatusCode     int               `json:"statusCode"`
	StatusText     string            `json:"statusText"`
	Latency        int64             `json:"latency"`
	StartedAt      string            `json:"startedAt"`
	EndedAt        string            `json:"endedAt"`
	RequestHeaders map[string]string `json:"requestHeaders,omitempty"`
}

// ── Event Queue ───────────────────────────────────────────────────────────────

const apioMaxQueue = 1000

var (
	apioQueue   = make([]apioEvent, 0, 256)
	apioQueueMu sync.Mutex
)

func apioEnqueue(ev apioEvent) {
	apioQueueMu.Lock()
	defer apioQueueMu.Unlock()
	if len(apioQueue) >= apioMaxQueue {
		apioQueue = apioQueue[1:] // drop oldest
	}
	apioQueue = append(apioQueue, ev)
}

func apioDrain() []apioEvent {
	apioQueueMu.Lock()
	defer apioQueueMu.Unlock()
	if len(apioQueue) == 0 {
		return nil
	}
	batch := make([]apioEvent, len(apioQueue))
	copy(batch, apioQueue)
	apioQueue = apioQueue[:0]
	return batch
}

// ── Background Sender ─────────────────────────────────────────────────────────

var apioSenderOnce sync.Once

func apioStartSender() {
	apioSenderOnce.Do(func() {
		if apioSDKToken == "" {
			fmt.Println("[api-monitor] ⚠️  APIO_SDK_TOKEN not set — monitoring disabled")
			return
		}
		fmt.Printf("[api-monitor] ✅ Sender active → %s\n", apioIngestURL)
		go func() {
			ticker := time.NewTicker(500 * time.Millisecond)
			defer ticker.Stop()
			client := &http.Client{Timeout: 5 * time.Second}
			for range ticker.C {
				apioDrainAndSend(client)
			}
		}()
	})
}

func apioDrainAndSend(client *http.Client) {
	batch := apioDrain()
	if len(batch) == 0 {
		return
	}
	payload, err := json.Marshal(map[string]interface{}{
		"sdkToken": apioSDKToken,
		"events":   batch,
	})
	if err != nil {
		return
	}
	resp, err := client.Post(apioIngestURL, "application/json", bytes.NewReader(payload))
	if err != nil {
		fmt.Printf("[api-monitor] ⚠️  Failed to send batch: %v\n", err)
		for _, ev := range batch {
			apioEnqueue(ev) // re-queue on failure
		}
		return
	}
	resp.Body.Close()
}

// ── Status-capturing ResponseWriter ───────────────────────────────────────────

type apioStatusWriter struct {
	http.ResponseWriter
	status int
	wrote  bool
}

func (sw *apioStatusWriter) WriteHeader(code int) {
	if !sw.wrote {
		sw.status = code
		sw.wrote = true
	}
	sw.ResponseWriter.WriteHeader(code)
}

func (sw *apioStatusWriter) Write(b []byte) (int, error) {
	if !sw.wrote {
		sw.status = http.StatusOK
		sw.wrote = true
	}
	return sw.ResponseWriter.Write(b)
}

// ── net/http Middleware ───────────────────────────────────────────────────────

// ApioMiddleware wraps any http.Handler (net/http, chi, gorilla/mux).
//
//	http.ListenAndServe(":8080", ApioMiddleware(yourMux))
func ApioMiddleware(next http.Handler) http.Handler {
	apioStartSender()
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		startedAt := time.Now()
		sw := &apioStatusWriter{ResponseWriter: w, status: http.StatusOK}

		url := "http://" + r.Host + r.RequestURI

		reqHeaders := make(map[string]string)
		for k, vs := range r.Header {
			reqHeaders[strings.ToLower(k)] = strings.Join(vs, ", ")
		}

		next.ServeHTTP(sw, r)

		latency := time.Since(startedAt).Milliseconds()
		statusText := "success"
		if sw.status >= 400 {
			statusText = "error"
		}

		apioEnqueue(apioEvent{
			Method:         r.Method,
			URL:            url,
			StatusCode:     sw.status,
			StatusText:     statusText,
			Latency:        latency,
			StartedAt:      startedAt.UTC().Format(time.RFC3339Nano),
			EndedAt:        time.Now().UTC().Format(time.RFC3339Nano),
			RequestHeaders: reqHeaders,
		})
	})
}

// ApioMiddlewareFunc is a chi-compatible middleware function.
//
//	r.Use(ApioMiddlewareFunc)
func ApioMiddlewareFunc(next http.Handler) http.Handler {
	return ApioMiddleware(next)
}

File: scripts/agents/apio_monitor.py
"""
apio_monitor.py
==================
Drop-in ASGI/WSGI middleware — paste into your Python project root.

Works with: FastAPI, Starlette, Flask, Django, Falcon — any Python web framework.
Zero new pip dependencies — uses only the standard library.

Setup:
  1. Copy this file to your project root
  2. Set env var: APIO_SDK_TOKEN=your_token
  3. Add middleware to your app (see examples below)
  4. Run your app normally

FastAPI / Starlette:
    from apio_monitor import ApiMonitorMiddleware
    app.add_middleware(ApiMonitorMiddleware)

Flask:
    from apio_monitor import WsgiApiMonitorMiddleware
    app.wsgi_app = WsgiApiMonitorMiddleware(app.wsgi_app)

Django (wsgi.py):
    from apio_monitor import WsgiApiMonitorMiddleware
    application = WsgiApiMonitorMiddleware(get_wsgi_application())

Config (env vars):
    APIO_SDK_TOKEN    — your service SDK token  (required)
    APIO_BACKEND_URL  — e.g. http://localhost:4000  (default)
"""

import os
import time
import json
import threading
import queue
from datetime import datetime, timezone
import urllib.request
import urllib.error

# ── Configuration ─────────────────────────────────────────────────────────────

SDK_TOKEN   = os.environ.get("APIO_SDK_TOKEN", "")
BACKEND_URL = os.environ.get("APIO_BACKEND_URL", "http://localhost:4000").rstrip("/")
INGEST_URL  = f"{BACKEND_URL}/api/v1/ingest"
BATCH_MS    = 500    # flush every 500 ms
MAX_QUEUE   = 1000   # drop oldest events when queue is full

# ── Event Queue ───────────────────────────────────────────────────────────────

_event_queue: queue.Queue = queue.Queue(maxsize=MAX_QUEUE)


def _enqueue(event: dict) -> None:
    try:
        _event_queue.put_nowait(event)
    except queue.Full:
        try:
            _event_queue.get_nowait()   # drop oldest
        except queue.Empty:
            pass
        _event_queue.put_nowait(event)


# ── Background Sender ─────────────────────────────────────────────────────────

def _drain_and_send() -> None:
    if not SDK_TOKEN:
        return
    batch = []
    try:
        while True:
            batch.append(_event_queue.get_nowait())
    except queue.Empty:
        pass
    if not batch:
        return

    payload = json.dumps({"sdkToken": SDK_TOKEN, "events": batch}).encode("utf-8")
    req = urllib.request.Request(
        INGEST_URL,
        data=payload,
        headers={"Content-Type": "application/json", "Content-Length": str(len(payload))},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=5):
            pass
    except Exception as exc:
        print(f"[api-monitor] ⚠️  Failed to send batch: {exc}")
        for ev in batch:
            _enqueue(ev)   # re-queue on failure


def _start_sender() -> None:
    def _loop():
        while True:
            time.sleep(BATCH_MS / 1000)
            _drain_and_send()

    t = threading.Thread(target=_loop, daemon=True, name="apio-sender")
    t.start()
    if SDK_TOKEN:
        print(f"[api-monitor] ✅ Sender active → {INGEST_URL}")
    else:
        print("[api-monitor] ⚠️  APIO_SDK_TOKEN not set — monitoring disabled")


_start_sender()


# ── ASGI Middleware — FastAPI / Starlette / Django 3+ ─────────────────────────

class ApiMonitorMiddleware:
    """
    ASGI middleware for FastAPI, Starlette, Django Channels.

    FastAPI:
        app.add_middleware(ApiMonitorMiddleware)

    Starlette:
        from starlette.middleware import Middleware
        app = Starlette(middleware=[Middleware(ApiMonitorMiddleware)])
    """

    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        started_at = datetime.now(timezone.utc)
        start_ns   = time.perf_counter_ns()

        method = scope.get("method", "GET")
        path   = scope.get("path", "/")
        query  = scope.get("query_string", b"").decode()
        server = scope.get("server", ("localhost", 80))
        host   = f"{server[0]}:{server[1]}"
        url    = f"http://{host}{path}" + (f"?{query}" if query else "")

        req_headers = {
            k.decode(): v.decode()
            for k, v in scope.get("headers", [])
        }

        status_code = [200]

        async def send_wrapper(message):
            if message["type"] == "http.response.start":
                status_code[0] = message.get("status", 200)
            await send(message)

        await self.app(scope, receive, send_wrapper)

        latency_ms = (time.perf_counter_ns() - start_ns) // 1_000_000
        ended_at   = datetime.now(timezone.utc)

        _enqueue({
            "method":         method,
            "url":            url,
            "statusCode":     status_code[0],
            "statusText":     "success" if status_code[0] < 400 else "error",
            "latency":        int(latency_ms),
            "startedAt":      started_at.isoformat().replace("+00:00", "Z"),
            "endedAt":        ended_at.isoformat().replace("+00:00", "Z"),
            "requestHeaders": req_headers,
        })


# ── WSGI Middleware — Flask / Django 2 / plain WSGI ──────────────────────────

class WsgiApiMonitorMiddleware:
    """
    WSGI middleware for Flask, Django (WSGI mode), Falcon.

    Flask:
        app.wsgi_app = WsgiApiMonitorMiddleware(app.wsgi_app)

    Django (wsgi.py):
        application = WsgiApiMonitorMiddleware(get_wsgi_application())
    """

    def __init__(self, wsgi_app):
        self.wsgi_app = wsgi_app

    def __call__(self, environ, start_response):
        started_at = datetime.now(timezone.utc)
        start_ns   = time.perf_counter_ns()

        method = environ.get("REQUEST_METHOD", "GET")
        path   = environ.get("PATH_INFO", "/")
        query  = environ.get("QUERY_STRING", "")
        host   = environ.get("HTTP_HOST", "localhost")
        url    = f"http://{host}{path}" + (f"?{query}" if query else "")

        req_headers = {
            k[5:].replace("_", "-").lower(): v
            for k, v in environ.items()
            if k.startswith("HTTP_")
        }

        status_holder = [200]

        def custom_start_response(status_str, headers, *args):
            status_holder[0] = int(status_str.split(" ", 1)[0])
            return start_response(status_str, headers, *args)

        result = self.wsgi_app(environ, custom_start_response)

        latency_ms = (time.perf_counter_ns() - start_ns) // 1_000_000
        ended_at   = datetime.now(timezone.utc)

        _enqueue({
            "method":         method,
            "url":            url,
            "statusCode":     status_holder[0],
            "statusText":     "success" if status_holder[0] < 400 else "error",
            "latency":        int(latency_ms),
            "startedAt":      started_at.isoformat().replace("+00:00", "Z"),
            "endedAt":        ended_at.isoformat().replace("+00:00", "Z"),
            "requestHeaders": req_headers,
        })

        return result
